### ISSUE TITLE
ENH: add ability to finetune / continue training

### ISSUE THREAD

**Issue by NickleDave** [vocalpy/vak#5]

**Comment by NickleDave** [comment]
Note to self: after thinking about this more, easier to add `checkpoint_path` to `train` and load from that path if present, instead of adding a whole new `--finetune` command to the `cli`.

Just need to clearly document

**Comment by kaiyaprovost** [comment]
I have added a module to address this issue which can be found in this pull: https://github.com/vocalpy/vak/pull/504

**Comment by NickleDave** [comment]
Great, thank you so much @kaiyaprovost for sharing the implementation from [your pre-print](https://www.biorxiv.org/content/10.1101/2022.02.24.481827v1).  

This will definitely help to see what changes we need to make.  
 
Like we discussed when we met, I feel like you and I are thinking about this the same way.  
But I also think we can achieve it without adding another module.

Instead (I think) we can just add a `ckpt_path` to the existing `vak.core.train` function.  It should default to `None`, in which case the function will behave as before, but when a `ckpt_path` is provided, we load pre-trained weights from that `ckpt_path`

Here's a to-do list for implementing in a new feature branch, with my explanation of the logic:  
- [ ] add a `ckpt_path` parameter to the existing `core.train` function that defaults to `None`
  - when `ckpt_path` is None, we do nothing and `train` functions as it did before
  - when `ckpt_path` is **not** None, then we try to load it -- this lets us continue training from the loaded checkpoint
- [ ] add a `ckpt_path` attribute to the `TrainConfig` class, that also defaults to None
- [ ] make sure that we pass that attribute into `vak.core.train` inside of `vak.cli.train`
- [ ] write additional unit tests that verify we can successfully load a `ckpt_path`
  - not immediately obvious to me how to test that it works, but we at least need a "smoke test" that makes sure nothing blows up when we try to do this
- [ ] add documentation explaining how to continue training / fine-tune models -- probably this should live in "how-tos" for now

I can help with writing unit tests, and I can write the docs if you'd like; just making note of all that here

This is just the list off the top of my head before looking in detail at your implementation. I will reply back with a little more detail once I've had a change to look at that.

**Comment by NickleDave** [comment]
Right so you have exactly the right idea in the way you've implemented it.

To add an optional `checkpoint_path` attribute to `TrainConfig` you can copy what `csv_path` does here:  https://github.com/vocalpy/vak/blob/5cb556ea4b3e740c5eb1f4916d3767d7082a9819/src/vak/config/train.py#L72  

so it would be  
```python
    checkpoint_path = attr.ib(
        converter=converters.optional(expanded_user_path),
        validator=validators.optional(is_a_file),
        default=None,
    )
```

**Comment by NickleDave** [comment]
As for loading the checkpoint itself inside `vak.core.train`, yes, we should do it exactly the way you have in your implementation:
https://github.com/kaiyaprovost/vak/blob/8d5cbeab8e7033c568afb5f7f79b73ffb596dae0/src/vak/core/train_checkpoint.py#L308

The only thing that will be different is we will need to add an `if` statement.

So the whole final code block should look like this, I think:  
```python
    for model_name, model in models_map.items():
        if checkpoint_path is not None:
            log_or_print(
                f"loading checkpoint for {model_name} from path: {checkpoint_path}",
                logger=logger,
                level="info",
            )
            model.load(checkpoint_path, device=device)

        results_model_root = results_path.joinpath(model_name)
        results_model_root.mkdir()
        ckpt_root = results_model_root.joinpath("checkpoints")
        ckpt_root.mkdir()
        log_or_print(f"training {model_name}", logger=logger, level="info")
        writer = tensorboard.get_summary_writer(
            log_dir=results_model_root, filename_suffix=model_name
        )
        model.summary_writer = writer
        model.fit(
            train_data=train_data,
            num_epochs=num_epochs,
            ckpt_root=ckpt_root,
            val_data=val_data,
            val_step=val_step,
            ckpt_step=ckpt_step,
            patience=patience,
            device=device,
        )
```

One possible issue here is that if it were ever the case that there multiple models then they would all try to load a single checkpoint path.   

I plan to remove that functionality anyway (it was never fully implemented).  So the `for model_name,  model` loop will go away and it won't matter in the long run.

**Comment by kaiyaprovost** [comment]
Hello Dave,
According to the notes from above I have started a new branch on my fork to address these changes. Please see the branch here: https://github.com/kaiyaprovost/vak/tree/modify-train-to-accept-checkpoints -- as per our email I believe this is how you wanted me to submit them rather than doing a full pull request. 

In addition to those changes adding checkpoint_path loading to each of the vak.*.train files, I have also taken the liberty to add similar functionality to load spect_scaler_path and labelmap_path, with no change to the default behavior (as they default to None). This was in my original implementation. I have submitted each of these separately.

**Comment by NickleDave** [comment]
>  Please see the branch here: https://github.com/kaiyaprovost/vak/tree/modify-train-to-accept-checkpoints

looks great, thank you

> as per our email I believe this is how you wanted me to submit them rather than doing a full pull request

ah whoops sorry for the confusion

In this case please **do** make a pull request and I will review it

**Comment by NickleDave** [comment]
> In addition to those changes adding checkpoint_path loading to each of the vak.*.train files, I have also taken the liberty to add similar functionality to load spect_scaler_path and labelmap_path, with no change to the default behavior (as they default to None). This was in my original implementation. I have submitted each of these separately.

Thank you, this is helpful to see.

I failed to notice you had added those additional parameters to the function in your first implementation.

It also had not dawned on me that one would want those parameters to continue training.

Of course you would want to--as you know, having already done it--and at first glance it looks like the way you implemented it will still work with our approach of making `vak.core.train` a multi-purpose function.
I don't think this will have any unintended consequences? Will come back to it with a fresh mind in the AM.

**Comment by NickleDave** [comment]
Thanks also for writing clear commit messages and using the abbreviations. Much appreciated.

In practice I usually end up making more messy and specific commits in a feature branch--e.g., "add `checkpoint_path` parameter to `vak.train.core`"-- and then squashing it down to a single commit with the abbreviation once I'm ready to merge.

For some of the commits it would be fine if you want to just go ahead and rewrite history to combine those now, as described here:
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
(that page gives a little more detail then the [`numpy` dev docs on the same thing](https://numpy.org/doc/1.14/dev/gitwash/development_workflow.html#rewriting-commit-history))

E.g. you could do an interactive rebase and make https://github.com/kaiyaprovost/vak/commit/ec5f85aaffcb31fc09fb96051754857272c90f97 a "fixup"  commit so that it just gets combined with https://github.com/kaiyaprovost/vak/commit/28fc2e1e7f0ebd13cc8de367cb85d997814046f6

This is one of those cases where it might be good to make a temporary copy of your repo before you attempt it, if you haven't done this before. But it's really helpful to be able to edit history on feature branches once you get the hang of it

**Comment by NickleDave** [comment]
Adding `spect_scaler_path` complicates things a bit because it can interact with the parameter `normalize_spectrograms`.

I'm writing down the use cases because trying to think about all the possible `if-else` combinations is making my brain hurt.
1. User trains a new model and **does** normalize spectrograms
  + we handle this already
2. User trains a new model and **does not** normalize
  + we handle this already, we know already in some cases we don't need to depending on spectrogram parameters, and this probably won't be necessary for future models that use batch normalization layers
3. User continues training and wants to load the previous `spect_scaler_path`
  + This makes sense for "picking up where you left off", e.g. training was interrupted or you want to test whether you could do better by training more on the same dataset
4. User continues training from a checkpoint but wants to create a new `spect_standardizer`, because they are training on a new dataset that has different statistics
  + This is the "fine-tuning" / transfer learning case
5. User continues training and **does not** normalize
  + e.g. because of spectrogram parameters or neural network layers as in use case 2

I think that's the Cartesian product of our options here :confused: 

A priori, before I look at the current code / PR again, this would look like (in lazy pseudo-Python)
```
if spect_scaler_path and normalize_spectrograms:  # use case 3
    spect_standardizer = joblib.load(spect_scaler_path)
elif spect_scaler_path and not normalize_spectrograms:
    raise ValueError(
        'spect_scaler_path provided but normalize_spectrograms was False, these options conflict'
)
elif normalize_spectrograms and spect_scaler_path is None:
    # instantiate a Standardizer and fit to data, this covers use case 1 and 4
else:
    spect_standardizer = None  # use case 2 and 5
```

To me this code snippet is showing the less-than-great way that `normalize_spectograms` interacts with `spect_scaler_path`.  What should happen if a person provides a `spect_scaler_path` but sets `normalize_spectrograms` to `False`? In the snippet I chose to raise an error to very loudly say "not sure what you're trying to do here but it seems conflicted?"

But annoyingly that would require someone using `spect_scaler_path` to also set `normalize_spectrograms` to `true` in their config.

The default is `true` so it's not the end of the world.  
But long-term I'm not sure we want the default to be `true`.  

An alternative would be to just ignore the setting of `normalize_spectrograms` when `spect_scaler_path is not None`.

I need to come back to this with fresh eyes but we definitely need some version of this logic to handle both parameters and all the uses cases outlined here.

**Comment by kaiyaprovost** [comment]
At the moment I have edited this to match your pseudo-code. In my opinion if someone is providing a spect_scaler_path then they have already normalized their training data and as such normalize_spectrograms should be set to true anyway for continuing a training job. However that is simply my logic. If that's the case I would suggest the code to look like this: 

```
    if spect_scaler_path is not None:
        log_or_print(
            f"loading spect scaler from path: {spect_scaler_path}",
            logger=logger,
            level="info")
        spect_standardizer = joblib.load(spect_scaler_path)
    # get transforms just before creating datasets with them
    elif normalize_spectrograms:
        log_or_print(
            f"no spect_scaler_path not provided, not loading",
            logger=logger,
            level="info",
        )
        spect_standardizer = None
        # we instantiate this transform here because we want to save it
        # and don't want to add more parameters to `transforms.split.get_defaults` function
        # and make too tight a coupling between this function and that one.
        # Trade off is that this is pretty verbose (even ignoring my comments)
        log_or_print("will normalize spectrograms", logger=logger, level="info")
        spect_standardizer = transforms.StandardizeSpect.fit_df(
            dataset_df, spect_key=spect_key
        )
        joblib.dump(spect_standardizer, results_path.joinpath("StandardizeSpect"))
    else: 
        spect_standardizer = None

```
